classes/services/NavigationMenuService.inc.php

```diff
@@ -0,0 +1,433 @@
+				$menuItemUrl = $dispatcher->url(
+					$request,
+					ROUTE_PAGE,
+					null,
+					'user',
+					'register',
+					null
+				);
+
+				$navigationMenuItem->setUrl($menuItemUrl);
+				break;
+			case NMI_TYPE_USER_LOGIN:
+				// Set navigationMenuItem type Display function
+				$navigationMenuItem->setIsDisplayed(!$isUserLoggedIn);
+
+				// Set navigationMenuItem type URL
+				$menuItemUrl = $dispatcher->url(
+					$request,
+					ROUTE_PAGE,
+					null,
+					'login',
+					null,
+					null
+				);
+
+				$navigationMenuItem->setUrl($menuItemUrl);
+				break;
+			case NMI_TYPE_CUSTOM:
+				// Set navigationMenuItem type Display function
+				$navigationMenuItem->setIsDisplayed(true);
+
+				if (!$navigationMenuItem->getUrl()) {
+					if ($navigationMenuItem->getPath()) {
+						// Set navigationMenuItem type URL
+						$menuItemUrl = $dispatcher->url(
+							$request,
+							ROUTE_PAGE,
+							null,
+							'navigationMenu',
+							'view',
+							$navigationMenuItem->getPath()
+						);
+
+						$navigationMenuItem->setUrl($menuItemUrl);
+					}
+				}
+
+				break;
+			default:
+				// Fire hook for determining display status of third-party types. Default: true
+				\HookRegistry::call('NavigationMenus::displayType', array(&$navigationMenuItem));
+
+
+		}
+	}
+
+	/**
+	 * Get a tree of NavigationMenuItems assigned to this menu
+	 *
+	 * @return array Hierarchical array of menu items
+	 */
+	public function getMenuTree(&$navigationMenu) {
+		$navigationMenuItemDao = \DAORegistry::getDAO('NavigationMenuItemDAO');
+		$items = $navigationMenuItemDao->getByMenuId($navigationMenu->getId())->toArray();
+		foreach($items as $item) {
+			$this->getDisplayStatus($item);
+		}
+
+		$navigationMenuItemAssignmentDao = \DAORegistry::getDAO('NavigationMenuItemAssignmentDAO');
+		$assignments = $navigationMenuItemAssignmentDao->getByMenuId($navigationMenu->getId())
+				->toArray();
+
+		for ($i = 0; $i < count($assignments); $i++) {
+			foreach($items as $item) {
+				if ($item->getId() === $assignments[$i]->getMenuItemId()) {
+					$assignments[$i]->setMenuItem($item);
+					break;
+				}
+			}
+		}
+
+		// Create an array of parent items and array of child items sorted by
+		// their parent id as the array key
+		$navigationMenu->menuTree = array();
+		$children = array();
+		foreach ($assignments as $assignment) {
+			if (!$assignment->getParentId()) {
+				$navigationMenu->menuTree[] = $assignment;
+			} else {
+				if (!isset($children[$assignment->getParentId()])) {
+					$children[$assignment->getParentId()] = array();
+				}
+				$children[$assignment->getParentId()][] = $assignment;
+			}
+		}
+
+		// Assign child items to parent in array
+		for ($i = 0; $i < count($navigationMenu->menuTree); $i++) {
+			$assignmentId = $navigationMenu->menuTree[$i]->getMenuItemId();
+			if (isset($children[$assignmentId])) {
+				$navigationMenu->menuTree[$i]->children = $children[$assignmentId];
+			}
+		}
+
+		// return $navigationMenu->menuTree;
```

Comment:
Should this be commented out? If so, remove it.

Reply:
OK